### ISSUE TITLE
Simplify using a custom TUser with DefaultUI

### ISSUE THREAD

**Issue by HaoK** [aspnet/Identity#1722]
This came up as a part of investigating the fix to https://github.com/aspnet/Identity/issues/1679

Today to add additional basic data to a custom user requires:
- overriding the Register/Manage pages, see https://github.com/aspnet/Identity/tree/dev/samples/IdentitySample.DefaultUI/Areas/Identity/Pages/Account, to hand craft the InputModel and update the actions.

```
   public class MyUser : IdentityUser {
       public string Name { get; set; }
       public int Age { get; set; }
   }
```

I think we can simplify this to only needing to define only the additional properties + the corresponding input model if we add a basic mechanism for generating the appropriate inputs and introduce a new abstraction in the UI package around updating/creating the actual TUser:

```C#
   public class MyUser : IdentityUser {
       public string Name { get; set; }
       public int Age { get; set; }
   }

   public class MyUserInputModel : RegisterInputModel, ManageInputModel {
            [Required]
            [DataType(DataType.Text)]
            [Display(Name = "Full name")]
            public string Name { get; set; }

            [Required]
            [Range(0,199, ErrorMessage = "Age must be between 0 and 199 years")]
            [Display(Name = "Age")]
            public int Age { get; set; }
   }

   public interface IUserFactory<TUser> {
       TUser Create(IUserData data); // Register calls this to create a new user
       void Update(TUser user, IUserData data); // Manage uses this to update the user
   }

   // We would provide a default implementation of this that uses reflection to work against properties, only users with truly custom POCOs would need to implement this.

    // We can add a way to generate the Register/Manage.cshtml inputs from the input models, like IInputGenerator.Generate(InputModel) to generate the form groups and give the ability to customize:
            <div class="form-group">
                <label asp-for="Input.Name"></label>
                <input asp-for="Input.Name" class="form-control" />
                <span asp-validation-for="Input.Name" class="text-danger"></span>
            </div>
            <div class="form-group">
                <label asp-for="Input.Age"></label>
                <input asp-for="Input.Age" class="form-control" />
                <span asp-validation-for="Input.Age" class="text-danger"></span>
            </div>
```

**Comment by HaoK** [comment]
Thoughts @DamianEdwards @danroth27 @javiercn @ajcvickers @blowdart @kichalla  @mkArtakMSFT ?

**Comment by mkArtakMSFT** [comment]
I've assigned this to rc1 milestone independent on the actual design decision here. We just know that there is some work to be done here.

**Comment by HaoK** [comment]
Chatted with @ajcvickers today he suggested we just make a small tweak to future proof the user factory without having to build everything, basically use an abstract base class as the dependency instead of an interface

`IUserFactory<TUser>` => abstract `UserFactoryBase<TUser>` 

This means we can add additional methods in 2.2 without having to deal with breaking changes.

**Comment by javiercn** [comment]
Moving this to 2.2

**Comment by javiercn** [comment]
For 2.1
* We will remove UserFactory<TUser>
* We will create users by calling Activator.CreateInstance<TUser>() and set the properties through store methods in our internal implementation.
* On the scaffolded code, we will rely on the properties available through IdentityUser.

This enables using the Default UI with Identity provided that:
* TUser has a parameter-less constructor or the user overrides the two views where users are created if they are used.
  * Register
  * ExternalLogin
* The underlying user store has support for emails or the user overrides the three views where emails are used.
  * Register
  * ExternalLogin
  * Manage/Index.

For 2.2

We will look into making the customizing a given user scenario easier. Some options that we've discussed are:
* Scaffold the relevant views when choosing a custom user.
  * Register
  * Manage/Index
  * ExternalLogins

**Comment by mkArtakMSFT** [comment]
@blowdart, is this something you want us to do?

**Comment by blowdart** [comment]
Yes, it would be incredibly useful. It's hard to discover right now

**Comment by blowdart** [comment]
@mkArtakMSFT we think the priority is more important than 3, because it's a long standing pain point. If Javier has not time for preview 2, can it be assigned to Identity, rather than Identity UI.

**Comment by mkArtakMSFT** [comment]
This definitely won't fit in @javiercn's schedule based on how much stuff he has on his plate.
@HaoK, is this something you can handle during 2.2?
/cc @ajcvickers

**Comment by HaoK** [comment]
Moving to preview3

**Comment by javiercn** [comment]
Sounds good to me. I think we need to discuss this before we jump into doing something. I'm not convinced on the approach and I haven't seen requests from customers for it.

**Comment by gulbanana** [comment]
I’ve got a customer request :) 

I’m integrating a proprietary ORM with ASP.NET Core and plugging it into Identity is *almost* easy. The one problem is that our entity classes aren’t parameterless - we use a different paradigm to EF where the “context” is not itself an app specific type but contains code-generated metadata binding app-specific types to generics. Internal details aside, it’s necessary for our business objects to have an IEntityContext at construction, like so:

`new ApplicationObject(ormHandle.CreateContext())`

In our asp.net integration these contexts can be acquired from DI, so if the default UI used DI to resolve a transient TUser that would be fine. Our IUserStore and UserManager<TUser> are also capable of creating contexts, so we can produce TUsers easily enough in FindById and friends. The problem is these Register and External pages which use Activator instead.

At the moment, we have an ugly workaround to work with the default identity ui: for TUser we use a wrapping class which lazily gets-or-creates the actual internal orm object using a context supplied by the IUserStore whenever it accesses a TUser. This is particularly annoying for users of our ORM because if they want to customise *our* representation of a user, they have to write something like
```
AddIdentityCore<ORMUser<ActualAppUser>>().AddORMStores();
[…]
await userManager.GetUser(User).GetOrCreateAsync(context)
```

In summary, we would appreciate a factory or DI resolution for TUser, as is already available everywhere except in some of this scaffolded UI.

**Comment by HaoK** [comment]
@gulbanana are you asking for something like a `IUserFactory` with a Create method that the UI would use instead?

**Comment by gulbanana** [comment]
That would work fine for my use case.

**Comment by DamianEdwards** [comment]
Can someone point me to the feedback from customers that are actually trying to create custom `IdentityUser` types that don't have constructorless parameters? I believe we can fix the templates to not instantly break when using a custom user type (just the code in `LoginPartial.cshtml` I believe). I'm interested as to whether we need to go this far though.

**Comment by DamianEdwards** [comment]
I see the one from @gulbanana above (obviously 😄) but are there more? Adding new factory classes and replumbing all the things to use that instead isn't trivial, and I'm concerned about assumptions being changed about who builds what in a minor release so want to be sure this is worthy of this kind of change, i.e. it benefits a good number of folks.

**Comment by HaoK** [comment]
This wasn't an issue until the UI so I doubt you will see too many issues about it as it wouldn't affect anyone who didn't switch.  We could do this transparently for all TUsers that derive from our built in IdentityUser types that have default constructors, the user factory would only be needed if you had a custom TUser that didn't derive from ours, which you'd just have to do something like this:

```
            services.AddDefaultIdentity<PocoUser>()
                .AddUserManager<UserManager<PocoUser>>()
                .AddDefaultUIUserFactory<PocoUser>((userName,email) =>
                    new PocoUser(userName, email);
```

**Comment by HaoK** [comment]
Discussed this with @ajcvickers and we decided it would be best to wait to tackle this in 3.0

**Comment by gulbanana** [comment]
I suspect there aren’t a lot of people with this use case, but I still think the change should be made (eventually. the concern about breaking changes is understandable). There’s a reason everything else uses DI instead of Activator!
